File: 20230204_Saturday/20230204.rb
# Leetcode: 567. Permutation in String.
# https://leetcode.com/problems/permutation-in-string/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 151 ms, faster than 82.00% of Ruby online submissions for Permutation in String.
# Memory Usage: 212.8 MB, less than 24.00% of Ruby online submissions for Permutation in String.
# 2023.02.04 Daily Challenge.
# @param {String} s1
# @param {String} s2
# @return {Boolean}
def check_inclusion(s1, s2)
  return false if s1.size > s2.size
  a =  count(s1)
  b =  count(s2[0...s1.size])
  return true if a == b
  (s1.size...s2.size).each do |j|
    prev = s2[j - s1.size].ord - ?a.ord
    curr = s2[j].ord - ?a.ord
    b[prev] -= 1
    b[curr] += 1
    return true if a == b
  end
  return false
end

def count(s)
  a = [0] * 26
  s.chars.each do |c|
    a[c.ord - ?a.ord] += 1
  end
  return a
end

File: 20230204_Saturday/20230204.go
// Leetcode: 567. Permutation in String.
// https://leetcode.com/problems/permutation-in-string/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 3 ms, faster than 87.72% of Go online submissions for Permutation in String.
// Memory Usage: 2.5 MB, less than 56.64% of Go online submissions for Permutation in String.
// 2023.02.04 Daily Challenge.
func checkInclusion(s1 string, s2 string) bool {
  l1,l2 := len(s1), len(s2)
  if l1 > l2 {
    return false
  }
  a,b := count(s1), count(s2[0:l1])
  if same(a,b){
    return true
  }
  for i,j := 0, l1; j < l2; i,j = i+1, j+1 {
    pr,cr := s2[i] - 'a', s2[j] - 'a'
    b[pr] -= 1
    b[cr] += 1
    if same(a,b) {
      return true
    }
  }
  return false
}

func count(s string) []int{
  ans := make([]int,26)
  for _,c := range s {
    ans[c - 'a'] += 1
  }
  return ans
}

func same(a,b []int) bool{
  for i,v := range a {
    if v != b[i] {
      return false
    }
  }
  return true
}
